source/common/upstream/upstream_impl.cc

```diff
@@ -987,7 +988,16 @@ void MainPrioritySetImpl::updateCrossPriorityHostMap(const HostVector& hosts_add
   }
 
   for (const auto& host : hosts_removed) {
-    mutable_cross_priority_host_map_->erase(addressToString(host->address()));
+    const auto host_address = addressToString(host->address());
+    const auto existing_host = mutable_cross_priority_host_map_->find(host_address);
+    if (existing_host != mutable_cross_priority_host_map_->end()) {
+      // An earlier priority added this host, which indicates that the host moved from
+      // a higher value priority to a lower value priority. We should not remove it from the
+      // cross priority map.
+      if (existing_host->second->priority() >= priority) {
```

Comment:
when will the `>` happens? Or should we only do the erasing when the priority is equal?

PS: I think this is a design problem of `corss_priority_map` which I can improve in future.

Reply:
changed, thanks for spotting. I think the thing I was thinking of is what I described in the comment 

```
// An earlier priority added this host, which indicates that the host moved from
      // a higher value priority to a lower value priority. We should not remove it from the
      // cross priority map.
```

but when the operators flipped it's more obvious that we only need to handle deletions when the priorities are equal. I've also updated the comment accordingly.

Regarding the design of cross_priority_map, I did think that because it's keyed on just the endpoint address, this seems to not capture differences of metadata of the endpoint but given that I was only focused on the priority part I wasn't sure if there was any other bugs here.

Reply:
Yeah, it initially is designed for lb to search a address quickly which could tolerate minor inaccuracy. But now, I need to re-think it.